keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -171,58 +183,37 @@ Windows related adjustments to default paths might be required.
-
-Long term and ideally, kube-proxy should be run as a DaemonSet on Windows.
+On Linux, kube-proxy is deployed as a DaemonSet in the kubeadm init phase. We will supply a DaemonSet that will run on Windows and use the wins API to launch kube-proxy.
 
 #### CNI plugin deployment
 
-On Linux, CNI plugins are deployed via kubectl and run as a DaemonSet. However, on Windows, CNI plugins need to run on the node, and cannot run in containers (again because Windows does not currently support privileged containers). Azure-CNI, win-bridge (compatible with kubenet), and Flannel all need the binary and config stored on the node.
+On Linux, CNI plugins are deployed via kubectl and run as a DaemonSet. We will supply a DaemonSet that will run Flannel configured in VXLAN/Overlay mode. This ideally can be upstreamed into the Flannel project.
 
-This proposal plans for FlannelD as the default option. Currently, FlannelD has to be started before the kube-proxy Windows service is started. FlannelD creates an HNS network on the Windows host, and kube-proxy will crash if it cannot find the network. This should be fixed in the scope of this project so that kube-proxy will wait until the network comes up. Therefore, kube proxy can be started at any time. 
+If the users wish to use a different plugin they can create a DaemonSet to be applied in place of the Flannel DaemonSet.
 
-However, if FlannelD is deployed in VXLAN (Overlay) mode, then we need to rewrite the KubeProxyConfiguration with the correct Overlay specific values, and kube-proxy will need to read this config again. This is not true for Host-Gateway (L2Bridge) mode. The script will have a flag that allows users to choose between the two networking modes.
+### Risks and Mitigations
 
-If the users wish to use a different plugin they will technically opt-out of the supported setup for kubeadm based Windows worker nodes in the Alpha release.
+**Risk**: Wins proxy introduces new security vector
 
-Long term, any CNI plugin should be supported for kubeadm based Windows worker nodes.
+The same functionality that allows us to now run privileged DaemonSets on Windows could be used maliciously to perform
+unwanted behavior on Windows nodes. This brings to Windows problems that already exist on Linux and now require the same
+mitigations, namely Pod Security Policies (PSP).
 
-### Risks and Mitigations
+*Mitigation*: Access to the wins named pipe can be restricted using a PSP that either disables
+`hostPath` volume mounts or restricts the paths that can be mounted. A sample PSP will be provided.
 
-**Risk**: Versioning of the wrapper script can become complicated
+**Risk**: Versioning of the script can become complicated
 
 Versioning of the script per-Kubernetes version can become a problem if a certain new version diverges in terms of download URLs, flags and configuration.
 
 *Mitigation*: Use git branches to version the script in the repository where it is hosted.
```

Comment:
we are not doing that, yet.
branches are much better than folders per-version, except that i'm not sure who has admin access to the repository.

WRT to maintenance if more people can merge changes in folders and nobody can create branches then folders is obviously the better solution.

Reply:
Yeah, again as I mentioned earlier I think this is significantly less of an issue with the new script being much smaller in scope, but I left this part in. I just wanted to make clear I don't want it to be wrapper anymore :smile: